pkg/spi/spidev_linux.go

```diff
@@ -0,0 +1,313 @@
+
+// Error implements the error interface.
+func (e ErrTxOverflow) Error() string {
+	return fmt.Sprintf("spi tx buffer overflow (%d > %d)", e.txLen, e.txMax)
+}
+
+// ErrRxOverflow is returned if the Transfer buffer is too large.
+type ErrRxOverflow struct {
+	rxLen, rxMax int
+}
+
+// Error implements the error interface.
+func (e ErrRxOverflow) Error() string {
+	return fmt.Sprintf("spi rx buffer overflow (%d > %d)", e.rxLen, e.rxMax)
+}
+
+// ErrBufferMismatch is returned if the rx and tx buffers do not have qual length.
+type ErrBufferMismatch struct {
+	txLen, rxLen int
+}
+
+// Error implements the error interface.
+func (e ErrBufferMismatch) Error() string {
+	return fmt.Sprintf("spi tx and rx buffers of unequal length (%d != %d)", e.txLen, e.rxLen)
+}
+
+// SPI wraps the Linux device driver and provides low-level SPI operations.
+type SPI struct {
+	f *os.File
+	// Used for mocking.
+	syscall func(trap, a1, a2 uintptr, a3 unsafe.Pointer) (r1, r2 uintptr, err unix.Errno)
+}
+
+// Open opens a new SPI device. dev is a filename such as "/dev/spidev0.0".
+// Remember to call Close() once done.
+func Open(dev string) (*SPI, error) {
+	f, err := os.OpenFile(dev, os.O_RDWR, 0)
+	if err != nil {
+		return nil, err
+	}
+	return &SPI{
+		f: f,
+		// a3 must be an unsafe.Pointer instead of a uintptr, otherwise
+		// we cannot mock out in the test without creating a race
+		// condition. See `go doc unsafe.Pointer`.
+		syscall: func(trap, a1, a2 uintptr, a3 unsafe.Pointer) (r1, r2 uintptr, err unix.Errno) {
+			return unix.Syscall(trap, a1, a2, uintptr(a3))
+		},
+	}, err
+}
+
+// Close closes the SPI device.
+func (s *SPI) Close() error {
+	return s.f.Close()
+}
+
+// Transfer performs multiple SPI reads and/or writes in a single function.
+// See the Transfer struct for details.
+func (s *SPI) Transfer(transfers []Transfer) error {
+	// Convert []Transfer to []iocTransfer.
+	it := make([]iocTransfer, len(transfers))
+	for i, t := range transfers {
+		it[i] = iocTransfer{
+			speedHz:        t.SpeedHz,
+			delayUSecs:     t.DelayUSecs,
+			bitsPerWord:    t.BitsPerWord,
+			txNBits:        t.TxNBits,
+			rxNBits:        t.RxNBits,
+			wordDelayUSecs: t.WordDelayUSecs,
+		}
+		if len(t.Tx) > maxTransferSize {
+			return ErrTxOverflow{len(t.Tx), maxTransferSize}
+		}
+		if len(t.Rx) > maxTransferSize {
+			return ErrRxOverflow{len(t.Rx), maxTransferSize}
+		}
+		if len(t.Tx) != 0 && len(t.Rx) != 0 && len(t.Tx) != len(t.Rx) {
+			return ErrBufferMismatch{len(t.Tx), len(t.Rx)}
+		}
+		if len(t.Tx) != 0 {
+			it[i].txBuf = uint64(uintptr(unsafe.Pointer(&transfers[i].Tx[0])))
+			it[i].length = uint32(len(t.Tx))
+		}
+		if len(t.Rx) != 0 {
+			it[i].rxBuf = uint64(uintptr(unsafe.Pointer(&transfers[i].Rx[0])))
+			it[i].length = uint32(len(t.Rx))
+		}
+		if transfers[i].CSChange {
+			it[i].csChange = 1
+		}
+	}
+
+	if _, _, err := s.syscall(unix.SYS_IOCTL, s.f.Fd(), uintptr(iocMessage(len(transfers))), unsafe.Pointer(&it[0])); err != 0 {
+		return err
```

Comment:
since this is just an errno, maybe wrap it in something nice? an os.SyscallError that adds some context, for example